libs/ledgerjs/packages/keyring-eth/src/DefaultKeyringEth.ts

```diff
@@ -1,46 +1,120 @@
+import { Transaction, ethers } from "ethers";
 import AppBinding from "@ledgerhq/hw-app-eth";
 import {
   EcdsaSignature,
   KeyringEth,
   GetAddressOptions,
   SignMessageOptions,
-  SignMessagePayload,
   SignTransactionOptions,
   GetAddressResult,
+  SignMessageMethod,
+  SignMessageType,
 } from "./KeyringEth";
+import { ContextResponse } from "@ledgerhq/context-module";
+import { ContextModule } from "@ledgerhq/context-module/lib/ContextModule";
 
 export class DefaultKeyringEth implements KeyringEth {
   private _appBinding: AppBinding;
+  private _contextModule: ContextModule;
 
-  constructor(transport: Transport) {
-    this._appBinding = new AppBinding(transport);
+  constructor(appBinding: AppBinding, contextModule: ContextModule) {
+    this._appBinding = appBinding;
+    this._contextModule = contextModule;
   }
 
   public async signTransaction(
-    _derivationPath: string,
-    _transaction: Transaction,
-    _options: SignTransactionOptions,
+    derivationPath: string,
+    transaction: Transaction,
+    options: SignTransactionOptions,
   ) {
-    // TODO: implement
-    return Promise.resolve({} as EcdsaSignature);
+    const challenge = await this._appBinding.getChallenge();
+
+    const contexts: ContextResponse[] = await this._contextModule.getContexts(transaction, {
+      challenge,
+      options,
+    });
+
+    for (const context of contexts) {
+      if (context.type === "error") {
+        // TODO: handle error here
+        continue;
+      }
+
+      await this._appBinding[context.type](context.payload);
+    }
+
+    const response = await this._appBinding.signTransaction(
+      derivationPath,
+      ethers.utils.serializeTransaction(transaction),
+      null, // context is already fetched by context-module
+    );
+
+    return {
+      r: `0x${response.r}`,
+      s: `0x${response.s}`,
+      v: parseInt(response.v, 16),
+    } as EcdsaSignature;
   }
 
-  public async signMessage(
-    _derivationPath: string,
-    _message: SignMessagePayload,
-    _options: SignMessageOptions,
-  ): Promise<EcdsaSignature> {
-    // TODO: implement
-    return Promise.resolve({} as EcdsaSignature);
+  public async signMessage<Method extends SignMessageMethod>(
+    derivationPath: string,
+    message: SignMessageType<Method>,
+    options: SignMessageOptions<Method>,
+  ) {
+    if (options.method === "personalSign") {
+      const personnalMessage = message as SignMessageType<"personalSign">;
+      const result = await this._appBinding.signPersonalMessage(
+        derivationPath,
+        Buffer.from(personnalMessage).toString("hex"),
+      );
+
+      return result as EcdsaSignature;
+    }
+
+    if (options.method === "eip712") {
+      const eip712Message = message as SignMessageType<"eip712">;
+      try {
+        const result = await this._appBinding.signEIP712Message(derivationPath, eip712Message);
+        return result as EcdsaSignature;
+      } catch (e) {
```

Comment:
I'm not sure we want to have this "any error" rule for the fallback. The introduction of this try/catch was initially for the nano S support where we would receive an INS_NOT_SUPPORTED error. 
Falling back all the time could be a solution to kinda attack this method by purposefully preventing clear signing.

Reply:
added a condition for `INS_NOT_SUPPORTED` error